Temporarily inhibit maintenance scans when node has heavy message load

vekterli opened the pull request:
@geirst please review.

If requests or responses from external sources are being constantly
processed as part of the distributor ticking, allow for up to _N_ ticks
to skip maintenance scanning, where _N_ is a configurable number.
This reduces the amount of CPU time spent on maintenance operations
when the node has a lot of incoming data to deal with.

Haven't seen much of a difference when testing locally, but I suspect that
my laptop is too scrawny to make the distributor the bottleneck when testing
end-to-end. Would like to get the factory performance tests' opinion on this.

vekterli committed:
Code cleanup, no functional changes

vekterli committed:
Allow temporarily inhibiting maintenance ops when under load

If requests or responses from external sources are being constantly
processed as part of the distributor tick, allow for up to N ticks
to skip maintenance scanning, where N is a configurable number.
This reduces the amount of CPU time spent on maintenance operations
when the node has a lot of incoming data to deal with.

baldersheim commented on storage/src/vespa/storage/config/stor-distributormanager.def:
```diff
@@ -236,4 +236,11 @@ use_weak_internal_read_consistency_for_client_gets bool default=false
+## maintenance scans from being done as part of the tick for up to N rounds of ticking.
+## This is to reduce the amount of CPU spent on ideal state calculations and bucket DB
+## accesses when the distributor is heavily loaded with feed operations.
+max_consecutively_inhibited_maintenance_ticks int default=20
```
What is the strange sign at the end of the line ?

vekterli replied:
That would be me forgetting to add a newline 😬

baldersheim commented:
The assumptions are correct, on the branch I have modifying the thread model. I manually just added 1 in 4 if more to do, and that improved performance significantly. This will make it even better.

vekterli commented:
Travis build was caught in the midst of our GCC 9 transition
```
Status: Downloaded newer image for vespaengine/vespa-build-centos7:latest
/source/travis/travis-build-full.sh: line 9: /etc/profile.d/enable-devtoolset-9.sh: No such file or directory
travis_time:end:0753fe36:start=1587136518407317906,finish=1587136658124084347,duration=139716766441,event=script
[0K[31;1mThe command "./travis/travis.sh" exited with 1.[0m
```

baldersheim commented:
Now it fails with 
storageframework_gtest_runner_app: /source/storageframework/src/vespa/storageframework/generic/thread/tickingthread.cpp:159: virtual void storage::framework::TickingThreadPoolImpl::start(storage::framework::ThreadPool&): Assertion `!_tickers.empty()' failed.

vekterli committed:
Remove test of error handling anti-pattern

Invariant violations should be assertions and not exceptions,
so no point in testing that this is handled gracefully.

geirst reviewed (approved):
👍
